wnd-charm porting on nvidia GPU

RutikaUbale opened the pull request:
Ported some algorithms of wnd-charm application on nvidia GPU

igg commented:
./configure  --enable-cuda
generates output like:

    ...
    checking whether to use CUDA library... yes
    configure: WARNING: Not using CUDA!
    ...
Nothing seems to be linked to cuda libs when running make.
Setup is cuda_6.5.14_mac_64 on Mac OS 10.10, installed (by default) in /usr/local/cuda. Sample progs and various nvidia-supplied cuda tests are passing.

The unexpected warning is not the result of testing for cuda presence, its from lines 42-45 in configure.ac:

    # check to see if CUDA been disabled
    if test "$cudaval" == "yes"; then
    	AC_MSG_WARN([Not using CUDA!])
    fi

That doesn't seem right - is this by intent?

Don't know how to progress further, since I can't seem to enable it.

lchandivade commented:
Ilya,

There seems to be some issue with the configure changes, we are fixing it. Meanwhile please use these steps

# autoreconf
# ./configure
# make

This should compile the code using the nvcc.

Thanks,
Lalit.

igg commented:
autoreconf generated a new configure that now triggers cuda compilation with no flags, but reports that cuda libraries will not be used.

Now some way into compilation, the error is:
```
nvcc  -I. -I..  -I/usr/local/cuda/include -L/usr/local/cuda/lib64 -lcuda -lcudart -lcufft -lcublas -c -o fft2.o fft2.cu
In file included from fft2.cu:30:
In file included from ./fft2.cuh:35:
In file included from ./ImageTransforms.h:36:
In file included from ./Tasks.h:35:
./unordered_map_dfn.h:3:11: fatal error: 'unordered_map' file not found
# include <unordered_map>
          ^
1 error generated.
make[2]: *** [fft2.o] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

There is some trickery in wndchrm's configure to use unordered_map if its available, or the old ordered map if its not.  This is one of the great pleasures of using C++ in its transitional phase.
The unordered_map_dfn.h file is supposed to take care of it and get a map of some sort defined.  Unfortunately it appears that nvcc is interpreting this differently.  I will take a look also and see if I can get to the bottom of it.

igg commented:
The problem stems from the fact that two different compilers are operating, but configure's checks only apply to the default c++ compiler.  Configure sets up the proper include path for unordered_map for the default c++ compiler (clang in my case), but this is not the correct setting for nvcc.  A stop-gap may be to check for nvcc in a #define, and include its version of unordered_map.  Without looking in detail, this fix probably belongs in unordered_map_dfn.h

lchandivade commented:
Would the application compile for CPU on this system?

igg commented:
Yes, OS X is the primary development platform, while the primary deployment platform (at least in our lab) is CentOS 6.  So both of these get very regular testing.
```
make distclean
./configure --disable-cuda
make
```
completes without error on this OS X system.

RutikaUbale commented:
Added follwing changes :- 
Removed unwanted header files from cuda modules.
Updated some modules for Memory leak issue.
Updated README.md file for cuda build.

It should build for OS X (64 bit ).

RutikaUbale commented:
Ported CombFirst4Moments algorithms to nvidia GPU.
Updated configuration to build windcharm GPU application for 32 bit cuda supported device.

igg commented:
Still can't compile it on OS X.  The error now is in linking:
Undefined symbols for architecture x86_64:
I tried with both gcc4.9 installed with homebrew as well as the standard clang-based dev tools provided by apple.

amolpunde commented:
We have tried several ways to compile wndchrm on Mac OSX.

To compile on Mac you need following things : 
CUDA 7.0 Toolkit.
Latest Eigen library(3.2.4)


Reason behind using CUDA 7.0 is : in CUDA 7.0 on Mac OS X, libc++ is supported with XCode 5.x. Command-line option -Xcompiler -stdlib=libstdc++ is no longer needed when invoking NVCC (Which is needed in CUDA 6.5, which results linking error). Instead, NVCC uses the default library that Clang chooses on Mac OS X.
If we use latest CUDA toolkit, we also need to get updated Eigen library 3.24.

Rest nothing need to be changed.

Thanks
Amol
